File: src/main/java/com/itmo/microservices/demo/payment/externalPaymentService/client/ExternalServiceClientImpl.java
package com.itmo.microservices.demo.payment.externalPaymentService.client;

import com.itmo.microservices.demo.payment.PaymentServiceConstants;
import com.itmo.microservices.demo.payment.externalPaymentService.client.model.ExternalServiceRequest;
import com.itmo.microservices.demo.payment.externalPaymentService.client.model.ExternalServiceResponse;
import com.itmo.microservices.demo.payment.externalPaymentService.client.model.ExternalServiceResponseStatus;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

@Slf4j
@RequiredArgsConstructor
@Component
public class ExternalServiceClientImpl implements ExternalServiceClient {

    private final RestTemplate externalPaymentServiceRestTemplate;

    @Override
    public ExternalServiceResponse executePayment(ExternalServiceRequest request) {
        var url = UriComponentsBuilder
                .fromPath(PaymentServiceConstants.EXTERNAL_SERVICE_TRANSACTIONS_URL)
                .build();

        try {
            var response = externalPaymentServiceRestTemplate
                    .postForObject(url.toUriString(), request, ExternalServiceResponse.class);
            if (response == null || response.getStatus() == ExternalServiceResponseStatus.FAILURE) {
                return null; // TODO:: replace with exception
            }

            return response;
        } catch (HttpClientErrorException | HttpServerErrorException e) {
            log.warn(e.getMessage());
            return null;
        }
    }
}

File: src/main/kotlin/com/itmo/microservices/demo/delivery/externalDeliveryService/client/ExternalServiceClientImpl.kt
package com.itmo.microservices.demo.delivery.externalDeliveryService.client

import com.itmo.microservices.demo.delivery.DeliveryServiceConstants
import com.itmo.microservices.demo.delivery.externalDeliveryService.client.model.ExternalServiceRequest
import com.itmo.microservices.demo.delivery.externalDeliveryService.client.model.ExternalServiceResponse
import org.springframework.stereotype.Component
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestTemplate
import org.springframework.web.client.postForObject
import org.springframework.web.client.getForObject
import org.springframework.web.util.UriComponentsBuilder

@Component
class ExternalServiceClientImpl(val externalDeliveryServiceRestTemplate: RestTemplate): ExternalServiceClient {

    override fun getDeliveryTransactionId(request: ExternalServiceRequest): ExternalServiceResponse? {
        val url = UriComponentsBuilder
            .fromPath(DeliveryServiceConstants.EXTERNAL_SERVICE_TRANSACTIONS_URL)
            .build()

        return try {
            externalDeliveryServiceRestTemplate
                .postForObject(url.toUriString(), request, ExternalServiceResponse::class)
        } catch (e: HttpClientErrorException) {
            null
        }
    }

    override fun getDeliveryParams(request: ExternalServiceRequest, transactionId: String): ExternalServiceResponse? {
        val url = UriComponentsBuilder
            .fromPath(DeliveryServiceConstants.EXTERNAL_SERVICE_TRANSACTIONS_URL + "/" + transactionId)
            .build()

        return try {
            externalDeliveryServiceRestTemplate
                .getForObject(url.toUriString(), request, ExternalServiceResponse::class)
        } catch (e: HttpClientErrorException) {
            null
        }
    }


}